attestation-service/attestation-service/src/policy_engine/opa/mod.rs

```diff
@@ -88,82 +81,56 @@ impl PolicyEngine for OPA {
                 hex::encode(hex)
             };
 
-            let policy_go = GoString {
-                p: policy.as_ptr() as *const c_char,
-                n: policy.len() as isize,
-            };
-
-            let reference = serde_json::json!({ "reference": reference_data_map }).to_string();
-
-            let reference_go = GoString {
-                p: reference.as_ptr() as *const c_char,
-                n: reference.len() as isize,
-            };
-
-            let input_go = GoString {
-                p: input.as_ptr() as *const c_char,
-                n: input.len() as isize,
-            };
-
-            // Call the function exported by cgo and process the returned decision
-            let decision_buf: *mut c_char =
-                unsafe { evaluateGo(policy_go, reference_go, input_go) };
-            let decision_str: &CStr = unsafe { CStr::from_ptr(decision_buf) };
-            let policy_res = decision_str.to_str()?.to_string();
-            debug!("OPA Evaluated: {}", policy_res);
-            if policy_res.starts_with("Error::") {
-                bail!("OPA verification failed: {policy_res}");
-            }
+            // Add policy as data
+            engine
+                .add_policy_from_file(policy_file_path)
+                .context("load policy")?;
 
-            // If a clear approval opinion is given in the evaluation report,
-            // the rejection information will be reflected in the evaluation failure return value.
-            let res_kv: Value = serde_json::from_str(&policy_res)?;
+            let reference_data_map = serde_json::to_string(&reference_data_map)?;
+            let reference_data_map =
+                regorus::Value::from_json_str(&format!("{{\"reference\":{reference_data_map}}}"))?;
+            engine
+                .add_data(reference_data_map)
+                .context("load reference data")?;
 
-            // only if there is a field named `allow` in the evaluation report and
-            // it is false, the evaluation fails. Otherwise the evaluation will be
-            // treated as succees.
-            let allow = res_kv
-                .get("allow")
-                .map(|a| a.as_bool())
-                .map(|a| a.unwrap_or(true))
-                .unwrap_or(true);
+            // Add TCB claims as input
+            engine.set_input_json(&input).context("set input")?;
 
+            let allow = engine.eval_bool_query("data.policy.allow".to_string(), false)?;
             if !allow {
-                bail!("TEE evidence does not pass policy {policy_id}, reason: {policy_res}");
-            } else {
-                let evaluation_result =
-                    serde_json::from_str(&policy_res).context("serialize OPA result")?;
-                res.insert(policy_id.to_owned(), (policy_hash, evaluation_result));
+                bail!("TEE evidence does not pass policy {policy_id}");
```

Comment:
Should we bail here or return an evaluation report that shows that the policy was false?

Reply:
The error handling here will finally be raised to a [RPC call error](https://github.com/confidential-containers/trustee/blob/main/attestation-service/attestation-service/src/bin/grpc/mod.rs#L195) for gRPC AS or [an internal error](https://github.com/confidential-containers/trustee/blob/main/attestation-service/attestation-service/src/bin/restful/mod.rs#L23-L30) for restful AS. If the policy all pass, an attestation token will be returned. So if we want a "failure report" to be returned, it would be practical to embed the reasons inside the error body returned to the client.

I did not find an API to gather the reason (like concrete field unmatched info) when a policy fails to check, thus only return the error log id. @anakrish Could you help with this?

Reply:
If more information is needed, then the policy must be written to provide additional information. For example, the policy in [rego tutorial](https://www.openpolicyagent.org/docs/latest/#2-try-opa-eval) not only has a rule for `allow`, but also has rules to capture the `violations`. 

To avoid executing two rules (`allow`, `violations`) separately, you could instead add a `violations` rule and rewrite the `allow` rule as below

```rego

violations := [ [k,v] | # Gather k, v and any other information.
   some k, v in input
   not judge_field(k, v)
]

allow := {
  "result": count(violations) == 0,
   "violations": violations
}

```
And use [eval_rule](https://docs.rs/regorus/0.1.5/regorus/struct.Engine.html#method.eval_rule) to evaluate `"data.policy.allow"`.

Another example is https://github.com/open-telemetry/weaver/blob/4401e3d2603121e6c21d0b4c6b47d8ef45151505/crates/weaver_checker/data/policies/otel_policies.rego#L13 where violations are gathered as json objects.

Reply:
I'm not sure we should use an Error for modelling control-flow of what seems to be valid code path: A policy doesn't pass validation. it's something that is expected and not really an internal error.

As far as I understand the semantics of this evaluate() fn seems to be:

- verify an input body against a set of ref values and a list of policies
- return a list of policy ids with their sha384 at eval time
- abort early on first failed validation

(maybe add a fn doc statement to make the behaviour clear obvious)

So, I think a signature like this could make sense?

```rust
    pub enum EvalOutcome {
       Passed(HashMap<String, PolicyDigest>),
       FailedWith { policy_id: String },
    }

    async fn evaluate(...) -> Result<EvalOutcome> {
    ...
    }
```

Reply:
> I'm not sure we should use an Error for modelling control-flow of what seems to be valid code path: A policy doesn't pass validation. it's something that is expected and not really an internal error.
> 
> As far as I understand the semantics of this evaluate() fn seems to be:
> 
> * verify an input body against a set of ref values and a list of policies
> * return a list of policy ids with their sha384 at eval time
> * abort early on first failed validation
> 
> (maybe add a fn doc statement to make the behaviour clear obvious)
> 
> So, I think a signature like this could make sense?
> 
> ```rust
>     pub enum EvalOutcome {
>        Passed(HashMap<String, PolicyDigest>),
>        FailedWith { policy_id: String },
>     }
> 
>     async fn evaluate(...) -> Result<EvalOutcome> {
>     ...
>     }
> ```

I vote for some docs upon the function for now.

Overall return enum makes sense to both tell the programmer that the function of this and the result. However a tricky thing is that if any policy does not pass, we should raise an error to the caller due to current logic.

The internal error thing is a workaround because all the error types inside attestation-service lib is `anyhow`, and we can easily map it to a HTTP error type [here](https://github.com/confidential-containers/trustee/blob/main/attestation-service/attestation-service/src/bin/restful/mod.rs#L18). Up to now it looks fine to me, as all errors we can just treat as `internal error`. Or, we can extend the successful response body format to support either a successful token, or a failure information in later PR.

Reply:
I agree it's a good idea to return an error to the caller, it's just the wrong error, IMO. "internal error" is reseved to signal the client that something is broken on the server, i.e. the client cannot do anything about it ([grpc](https://grpc.github.io/grpc/core/md_doc_statuscodes.html), [rest](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500) semantics)

> Internal errors. This means that some invariants expected by the underlying system have been broken. This error code is reserved for serious errors.

> 500 Internal Server Error server error response code indicates that the server encountered an unexpected condition that prevented it from fulfilling the request.

if a policy doesn't pass or the client calls the with `policy_ids=vec!(["does-not-exist"])`, it's not an internal error, it's a user error.

Reply:
Right. we need to define a more generalized 200 response, with either token or error info.

Reply:
we can still return an error, but we should maybe discern:
- user errors (e.g. policy doesn't exist => 400)
- internal errors (e.g. cannot perform ls in /opt/... => 500)
- result of the validation (the upper-layer logic, probably in get_token() that is calling the fn should decide whether fail-to-pass is good or bad, so it could be either 400 or 200 for a client)